### PR TITLE
new audit optimizations

### PR DESCRIPTION
**Pull request by psorinionut**
- Improved user experience by making the subtract payment operation based on energy automatic
- Improved and simplified payment computation algorithms
- Add/remove farms endpoints callable by admins
- Updated checks and verifications

### CODE REVIEW

File: farm-boosted-rewards-subscriber/src/service.rs [hunk 1]
```diff
@@ -33,51 +34,73 @@ pub trait ServiceModule:
...
-        require!(
-            service_index == SubscriptionUserType::Normal as usize
-                || service_index == SubscriptionUserType::Premium as usize,
-            "Invalid service index"
-        );
-        let is_premium_service = service_index == SubscriptionUserType::Premium as usize;
+        let current_epoch = self.blockchain().get_block_epoch();
+        let payment_recurrency = EPOCHS_IN_WEEK;
+        let standard_service_index = SubscriptionUserType::Normal as usize;
+        let premium_service_index = SubscriptionUserType::Premium as usize;
```
Comment (dorin-iancu) [inline farm-boosted-rewards-subscriber/src/service.rs:42]: Bad practice to use enums as values. Put this into a method of `SubscriptionUserType`, like `get_service_index(&self)`

File: farm-boosted-rewards-subscriber/src/service.rs [hunk 2]
```diff
@@ -33,51 +34,73 @@ pub trait ServiceModule:
...
+                user_last_payment_mapper.get()
+            };
+
+            if current_epoch < user_last_payment.epoch + payment_recurrency
+                && user_last_payment.epoch > 0
+            {
+                continue;
+            }
+
+            let user = opt_user_address.unwrap();
```
Comment (dorin-iancu) [inline farm-boosted-rewards-subscriber/src/service.rs:69]:
```suggestion
            let user = unsafe { opt_user_address.unwrap_unchecked() };
```
You already checked for `None`.

File: farm-boosted-rewards-subscriber/src/subscriber_config.rs [hunk 3]
```diff
@@ -1,14 +1,28 @@
...
 pub const EPOCHS_IN_WEEK: u64 = 7;
 
+#[derive(Default, TypeAbi, TopEncode, TopDecode)]
+pub struct UserLastPayment {
+    pub service_index: usize,
+    pub epoch: Epoch,
+}
+
+impl UserLastPayment {
+    pub fn new(service_index: usize, epoch: Epoch) -> Self {
```
Comment (dorin-iancu) [inline farm-boosted-rewards-subscriber/src/subscriber_config.rs:18]: I'd remove this method, just initialize in-line if there are no additional conditions for the arguments.

File: farm-boosted-rewards-subscriber/src/service.rs (hunk 2 above)
Comment (psorinionut) [reply farm-boosted-rewards-subscriber/src/service.rs:69]: Done.

File: farm-boosted-rewards-subscriber/src/subscriber_config.rs (hunk 3 above)
Comment (psorinionut) [reply farm-boosted-rewards-subscriber/src/subscriber_config.rs:18]: Done.

File: farm-boosted-rewards-subscriber/src/service.rs (hunk 1 above)
Comment (psorinionut) [reply farm-boosted-rewards-subscriber/src/service.rs:42]: I know we usually avoid coding like this, but here I wanted to make sure it's clear we can only have these 2 options. I've changed the enum to two different constants. For how to code looks like now, I don't think a new method would be of much help.

**Review by dorin-iancu — APPROVED** [review]

**Pull request merged**